### PR TITLE
Added Randomized Leaky ReLU operator to Keras

### PR DESCRIPTION
**Pull request by amitsrivastava78**
Added advance activation function Randomized Leaky ReLU to keras layer

### CODE REVIEW

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 1]
```diff
@@ -333,3 +333,63 @@ def get_config(self):
...
+    if lower > upper :
+      raise ValueError('Upper bound for RandomizedRectifierLayer needs '
+                       'to be higher than lower bound. ')
+
+    self.lower = K.cast_to_floatx(lower)
+    self.upper = K.cast_to_floatx(upper)
+    self.seed = K.cast_to_floatx(seed)
+    self.alpha = K.random_uniform_variable(shape=self.lower.shape,low=self.lower,
+                                           high=self.upper,seed=self.seed,dtype=float)
+    self.alpha = self.alpha.numpy()
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: Is this going to be v1 compatible?

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 2]
```diff
@@ -333,3 +333,63 @@ def get_config(self):
...
+    super(RReLU, self).__init__(**kwargs)
+    if lower < 0 or upper < 0  :
+      raise ValueError('lower/upper bound for RandomizedRectifierLayer needs '
+                       'to be > 0. ')
+    if lower > upper :
+      raise ValueError('Upper bound for RandomizedRectifierLayer needs '
+                       'to be higher than lower bound. ')
+
+    self.lower = K.cast_to_floatx(lower)
+    self.upper = K.cast_to_floatx(upper)
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: Can you check shape equivalence?

File: tensorflow/python/keras/layers/advanced_activations_test.py [hunk 3]
```diff
@@ -88,6 +88,30 @@ def test_relu_with_invalid_arg(self):
             kwargs={'negative_slope': -2},
             input_shape=(2, 3, 4))
 
+  def test_rrelu_with_invalid_arg(self):
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations_test.py]: This might needs testing for main case, i.e., create a RRelu and check the output in some way.

**Review by tanzhenyu — CHANGES_REQUESTED** [review]

File: tensorflow/python/keras/layers/advanced_activations.py (hunk 1 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: This is V2 spec, i have added few more modification to make it as per the spec

File: tensorflow/python/keras/layers/advanced_activations_test.py (hunk 3 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations_test.py]: Added one more test cases as requested by you and found a method mentioned in the spec to test it.This follows similar to the dropout method.

File: tensorflow/python/keras/layers/advanced_activations.py (hunk 2 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: These are scalars and type check might be redundant.

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 4]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
   @tf_utils.shape_type_conversion
   def compute_output_shape(self, input_shape):
     return input_shape
+
+@keras_export('keras.layers.RReLU')
+class RReLU(Layer):
+  """Randomized Leaky Rectified Liner Unit function
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: Typo

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 5]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
...
+    def rrelu_inputs_testing():
+       with ops.name_scope("Random_Relu_Testing"):
+            alpha = (self.lower+ self.upper) / 2
+            return K.relu(inputs,alpha=alpha)
+
+    def rrelu_input_training():
+       return K.relu(inputs,alpha=self.alpha)
+
+    with ops.name_scope(self.name, "random-relu", [inputs]):
+       output = tf_utils.smart_cond(self.training,
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: Seems like training should be a variable or tensor here, otherwise why use tf.cond? I would guess K.learning_phase is a good start.

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 6]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
...
+  Output shape:
+      Same shape as the input.
+
+  Arguments:
+      lower: lower bound of the uniform distribution.
+      upper: upper bound of the uniform distribution.
+      seed: Used to create random seeds.
+  """
+
+  def __init__(self, lower=0.125, upper=0.333, seed=None, training=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Sure , i have done and updated the same

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 7]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
...
+    self.seed =  seed
+    self.training = training
+    if self.training==True:
+       self.alpha = K.random_range(low=lower, high=upper)
+
+  def call(self, inputs):
+
+    def rrelu_inputs_testing():
+       with ops.name_scope("Random_Relu_Testing"):
+            alpha = (self.lower+ self.upper) / 2
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Yes you are right, i have rectified the same and updated the code

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 8]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
...
+    self.upper = upper
+    self.seed =  seed
+    self.training = training
+    if self.training==True:
+       self.alpha = K.random_range(low=lower, high=upper)
+
+  def call(self, inputs):
+
+    def rrelu_inputs_testing():
+       with ops.name_scope("Random_Relu_Testing"):
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Rectified this part

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 9]
```diff
@@ -333,3 +334,74 @@ def get_config(self):
...
+            alpha = (self.lower+ self.upper) / 2
+            return K.relu(inputs,alpha=alpha)
+
+    def rrelu_input_training():
+       return K.relu(inputs,alpha=self.alpha)
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Rectified this part

File: tensorflow/python/keras/layers/advanced_activations.py (hunk 4 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Rectified this part

File: tensorflow/python/keras/layers/advanced_activations.py (hunk 5 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: Yes you are write this is smarter way to do it.

Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: @tanzhenyu , thanks alot for the review it has helped alot to improve the quality of the code.I have made all the changes as requested by you,pls check

**Comment by amitsrivastava78** [comment]
@tanzhenyu , can you pls have a look, i have updated everything as per your review comments

**Comment by hgadig** [comment]
> @tanzhenyu @hgadig , can you pls have a look, i have updated everything as per your review comments

Thanks. But can you please rebase as I can see there are branch conflicts. Rebase would solve this.

**Comment by amitsrivastava78** [comment]
@hgadig & @tanzhenyu , i have rebased, pylint and tested the code, kindly approve and merge

**Comment by hgadig** [comment]
> @hgadig & @tanzhenyu , i have rebased, pylint and tested the code, kindly approve and merge

@tanzhenyu  Could you please take a look and approve this.

File: tensorflow/python/keras/backend.py [hunk 10]
```diff
@@ -1151,6 +1151,21 @@ def identity(x, name=None):
   return array_ops.identity(x, name=name)
 
 
+def random_range(low=0, high=1, size=None):
+  """Returns a random variable withn the range including low,
```
Comment (tanzhenyu) [inline tensorflow/python/keras/backend.py]: 'within'

Comment (tanzhenyu) [inline tensorflow/python/keras/backend.py]: of [low, high)

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 11]
```diff
@@ -333,3 +334,73 @@ def get_config(self):
...
+      (tuple of integers, does not include the samples axis)
+      when using this layer as the first layer in a model.
+
+  Output shape:
+      Same shape as the input.
+
+  Arguments:
+      lower: lower bound of the uniform distribution.
+      upper: upper bound of the uniform distribution.
+      seed: Used to create random seeds.
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: I don't see where this seed is being used.

**Comment by amitsrivastava78** [comment]
@tanzhenyu & @hgadig , i have fixed all the issues pointed out, again did the pylint, (even spell check this time) and tested the code

File: tensorflow/python/keras/layers/advanced_activations.py [hunk 12]
```diff
@@ -333,3 +334,70 @@ def get_config(self):
...
+    if lower < 0 or upper < 0:
+      raise ValueError('lower/upper bound for RandomizedRectifierLayer needs '
+                       'to be > 0. ')
+    if lower > upper:
+      raise ValueError('Upper bound for RandomizedRectifierLayer needs '
+                       'to be higher than lower bound. ')
+
+    self.lower = lower
+    self.upper = upper
+    self.alpha = K.random_range(low=lower, high=upper)
```
Comment (tanzhenyu) [inline tensorflow/python/keras/layers/advanced_activations.py]: Without specifically using "K.cast_to_floatx", will this be able to support different floating point precision?

**Comment by tanzhenyu** [comment]
Also I think it makes more sense to place it in tensorflow addons, instead of tf.keras.
Please see: https://github.com/tensorflow/addons

(In general we only maintain that in core once the api and implementation is near finalized)

File: tensorflow/python/keras/backend.py [hunk 13]
```diff
@@ -1151,6 +1151,21 @@ def identity(x, name=None):
   return array_ops.identity(x, name=name)
 
 
+def random_range(low=0, high=1, size=None):
```
[the remark this answers is not in the document]
Comment (amitsrivastava78) [reply tensorflow/python/keras/backend.py]: @tanzhenyu , please find my answers inline
1. Actually, this is what initailly i implemented and then i found that backend is the one which is calling all the numpy interface, so it felt logical to me to put it as backend interface.
2. This behavior is similar to the Drop out, and yes it is as per the paper implmentation, you can refer to section 2.4 of the same.

File: tensorflow/python/keras/layers/advanced_activations.py (hunk 12 above)
Comment (amitsrivastava78) [reply tensorflow/python/keras/layers/advanced_activations.py]: At present all my TCs are passing , if you feel K.cast_to_floatx is better way to go i am up for change.

**Comment by amitsrivastava78** [comment]
@tanzhenyu , this is advance activation function which is implmented by many frameworks including pytorch, you can refer to it and see that this activation function is put along with other activation functions, i feel this folder might be the correct place for the same, let me know what you feel.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @tanzhenyu: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @tanzhenyu: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tanzhenyu** [comment]
Same comment as before, we could put them in add-on.

**Comment by amitsrivastava78** [comment]
@tanzhenyu, thanks for the review, as suggested i am closing the PR.

Regards 
Amit
